iquip/apps/builder.py

```diff
@@ -0,0 +1,419 @@
+
+    Attributes:
+        name: The argument name.
+        nameLabel: The label for showing the argument name.
+    """
+
+    def __init__(self, name: str, parent: Optional[QWidget] = None):
+        """Extended.
+        
+        Args:
+            name: See the attributes section in _BaseEntry.
+        """
+        super().__init__(parent=parent)
+        self.name = name
+        # widgets
+        self.nameLabel = QLabel(name, self)
+        # layout
+        self.layout = QHBoxLayout(self)
+        self.layout.addWidget(self.nameLabel)
+        self.setLayout(self.layout)
+
+    def value(self) -> Any:
+        """Returns the entered or selected value.
+        
+        This must be overridden in the subclass.
+        """
+        raise NotImplementedError
+
+
+class _BooleanEntry(_BaseEntry):
+    """Entry class for a boolean value."""
+
+    def __init__(self, name: str, default: bool = False, parent: Optional[QWidget] = None):
+        """Extended.
+        
+        Args:
+            default: The default value. If it does not exist, it is set to False.
+        """
+        super().__init__(name, parent=parent)
+        # widgets
+        self.checkBox = QCheckBox(self)
+        self.checkBox.setChecked(default)
+        # layout
+        self.layout.addWidget(self.checkBox)
+
+    def value(self) -> bool:
+        """Overridden.
+        
+        Returns the status of the checkBox.
+        """
+        return self.checkBox.isChecked()
+
+
+class _EnumerationEntry(_BaseEntry):
+    """Entry class for an enumeration value."""
+
+    def __init__(
+        self,
+        name: str,
+        choices: List[str],
+        default: Optional[str] = None,
+        parent: Optional[QWidget] = None
+    ):
+        """Extended.
+        
+        Args:
+            choices: The pre-defined candidates.
+            default: The default value. If it does not exist, it is set to the first candidate.
+        """
+        super().__init__(name, parent=parent)
+        # widgets
+        self.comboBox = QComboBox(self)
+        for choice in choices:
+            self.comboBox.addItem(choice)
+        if choices:
+            self.comboBox.setCurrentText(choices[0] if default is None else default)
+        # layout
+        self.layout.addWidget(self.comboBox)
+
+    def value(self) -> str:
+        """Overridden.
+        
+        Returns the value of the comboBox.
+        """
+        return self.comboBox.currentText()
+
+
+class _NumberEntry(_BaseEntry):
+    """Entry class for a number value.
+    
+    Attributes:
+        scale: The scale factor that actually applies.
+        type: The type of the value. If "int", value() returns an integer value.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        unit: str,
+        scale: float,
+        step: float,
+        min: Optional[float],  # pylint: disable=redefined-builtin
+        max: Optional[float],  # pylint: disable=redefined-builtin
```

Comment:
Let's not redefine the built-in identifiers. You could use `minimum` or `min_`.

Reply:
I hope that, too... but I want to use the original parameter name which is returned by the proxy server.

I came up with three solutions.
1. Modify `initArgsEntry()` to replace `min` and `max` to other names of `argInfo`.
For this, we should handle a `NumberValue` separately.
2. Modify the artiq proxy to change arguments name.
Similar to the above solution, we should handle separately.
3. Receive the arguments as a dict not separately.
4. Keep it as it is.

I think solution 3 is best, how about you?

Reply:
I also think 3 is the best. Then you can keep the docstring like:
```python
"""...

Args:
    argInfo: ... handled arguments are:
        ...
        min: ... (default=0)
        ...
        type: 
"""
```

Reply:
Good! Then, is it okay to keep the key names such as `min`, `type`, and `ndecimals`?

Reply:
Yes, it is okay since they are just strings.